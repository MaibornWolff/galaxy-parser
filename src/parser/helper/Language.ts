import CSharp from "tree-sitter-c-sharp";
import CPlusPlus from "tree-sitter-cpp";
import GO from "tree-sitter-go";
import Java from "tree-sitter-java";
import JavaScript from "tree-sitter-javascript";
import Kotlin from "tree-sitter-kotlin";
import PHP from "tree-sitter-php";
import Python from "tree-sitter-python";
import TypeScript from "tree-sitter-typescript";
import Ruby from "tree-sitter-ruby";
import Rust from "tree-sitter-rust";
import Bash from "tree-sitter-bash";
import C from "tree-sitter-c";
import JSON from "tree-sitter-json";
import { ConstantTwoWayMap } from "./ConstantTwoWayMap";
import { Configuration } from "../Configuration";
import { getFileExtension, lookupLowerCase, replaceForwardWithBackwardSlashes } from "./Helper";
import path from "path";

/**
 * Enum of all supported programming languages.
 */
export const enum Language {
    CSharp,
    CPlusPlus,
    Go,
    Java,
    JavaScript,
    Kotlin,
    PHP,
    TypeScript,
    TSX,
    Python,
    Ruby,
    Rust,
    Bash,
    C,
    JSON,
}

/**
 * Maps back and forth between language and an abbreviation as string value.
 * The string value corresponding to the language is equivalent to the abbreviation used in the node mappings file.
 *
 * Note that this is not necessarily identical to the file extension of a file of the corresponding language,
 * as there can be multiple file extensions for one language. Use {@link fileExtensionToLanguage}
 * or {@link fileExtensionToGrammar} for mapping file extensions.
 */
export const languageToAbbreviation = new ConstantTwoWayMap<Language, string>(
    new Map([
        [Language.CSharp, "cs"],
        [Language.CPlusPlus, "cpp"],
        [Language.Go, "go"],
        [Language.Java, "java"],
        [Language.JavaScript, "js"],
        [Language.Kotlin, "kt"],
        [Language.PHP, "php"],
        [Language.TypeScript, "ts"],
        [Language.TSX, "tsx"],
        [Language.Python, "py"],
        [Language.Ruby, "rb"],
        [Language.Rust, "rs"],
        [Language.Bash, "sh"],
        [Language.C, "c"],
        [Language.JSON, "json"],
    ])
);

/**
 * Maps from a language to the corresponding tree-sitter grammar.
 */
export const languageToGrammar = new Map([
    [Language.CSharp, CSharp],
    [Language.CPlusPlus, CPlusPlus],
    [Language.Go, GO],
    [Language.Java, Java],
    [Language.JavaScript, JavaScript],
    [Language.Kotlin, Kotlin],
    [Language.PHP, PHP.php],
    [Language.TypeScript, TypeScript.typescript],
    [Language.TSX, TypeScript.tsx],
    [Language.Python, Python],
    [Language.Ruby, Ruby],
    [Language.Rust, Rust],
    [Language.Bash, Bash],
    [Language.C, C],
    [Language.JSON, JSON],
]);

/**
 * Maps supported file extensions to the corresponding programming languages.
 * In lower case. This list is for file extensions which are not case-sensitive.
 */
const fileExtensionToLanguage = new Map([
    ["cs", Language.CSharp],
    ["cpp", Language.CPlusPlus],
    ["cp", Language.CPlusPlus],
    ["cxx", Language.CPlusPlus],
    ["cc", Language.CPlusPlus],
    ["hpp", Language.CPlusPlus],
    ["hxx", Language.CPlusPlus],
    ["hh", Language.CPlusPlus],
    ["go", Language.Go],
    ["java", Language.Java],
    ["js", Language.JavaScript],
    ["kt", Language.Kotlin],
    ["php", Language.PHP],
    ["ts", Language.TypeScript],
    ["tsx", Language.TSX],
    ["py", Language.Python],
    ["rb", Language.Ruby],
    ["rs", Language.Rust],
    ["sh", Language.Bash],
    ["json", Language.JSON],
]);

/**
 * Maps supported file extensions to the corresponding programming languages.
 * For case-sensitive file extensions.
 */
const caseSensitiveFileExtensionToLanguage = new Map([
    ["c", Language.C],
    ["C", Language.CPlusPlus],
    ["H", Language.CPlusPlus], // lowercase .h has a special treatment
]);

/**
 * Estimates the language of a file based upon the file extension and file path.
 * @param filePath Path to the file, including the file extension.
 * @param config Configuration to apply.
 * @param pathModule ONLY FOR TESTING PURPOSES: overrides the platform-specific path module.
 */
export function assumeLanguageFromFilePath(
    filePath: string,
    config: Configuration,
    pathModule = path,
) {
    const fileExtension: string = getFileExtension(filePath);

    if (fileExtension === "h") {
        if (shouldHBeParsedAsC(filePath, config, pathModule)) {
            return Language.C;
        }
        return Language.CPlusPlus;
    }

    const resultCaseSensitive = caseSensitiveFileExtensionToLanguage.get(fileExtension);
    if (resultCaseSensitive !== undefined) {
        return resultCaseSensitive;
    }
    return lookupLowerCase(fileExtensionToLanguage, fileExtension);
}

/**
 * Handling of the parse .h as C option.
 */
function shouldHBeParsedAsC(filePath: string, config: Configuration, pathModule = path): boolean {
    if (config.parseAllHAsC) {
        return true;
    }
    if (config.parseSomeHAsC.size > 0) {
        // Use the path relative to the sources path to avoid the unintuitive behaviour
        // that higher-level folders are evaluated for this:
        const relativePath = pathModule.relative(config.sourcesPath, filePath);
        const backwardSlashRelpath = replaceForwardWithBackwardSlashes(relativePath);
        const relpathSplitted = backwardSlashRelpath.split("\\");
        for (const pathElement of relpathSplitted) {
            if (config.parseSomeHAsC.has(pathElement)) {
                return true;
            }
        }
    }
    return false;
}

/**
 * Maps supported file extensions to the corresponding language grammar.
 * @param fileExtension The file extension to map.
 * @return The corresponding grammar if the file extension is supported, undefined otherwise.
 */
export function fileExtensionToGrammar(fileExtension: string) {
    const language = fileExtensionToLanguage.get(fileExtension);
    if (language !== undefined) {
        return languageToGrammar.get(language);
    } else {
        return undefined;
    }
}
