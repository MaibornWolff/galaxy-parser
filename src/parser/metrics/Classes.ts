import { QueryBuilder } from "../queries/QueryBuilder";
import { NodeTypeCategory, NodeTypeConfig } from "../helper/Model";
import { FileMetric, Metric, MetricResult, ParsedFile } from "./Metric";
import { debuglog, DebugLoggerFunction } from "node:util";
import { QueryMatch } from "tree-sitter";
import {
    QueryStatementInterface,
    SimpleLanguageSpecificQueryStatement,
} from "../queries/QueryStatements";
import { getQueryStatementsByCategories } from "../helper/Helper";

let dlog: DebugLoggerFunction = debuglog("metric-gardener", (logger) => {
    dlog = logger;
});

export class Classes implements Metric {
    private readonly statementsSuperSet: QueryStatementInterface[] = [];

    private nodeTypeCategories = new Set([
        NodeTypeCategory.ClassDefinition,
        NodeTypeCategory.EnumDefinition,
        NodeTypeCategory.StructDefinition,
        NodeTypeCategory.RecordDefinition,
        NodeTypeCategory.UnionDefinition,
        NodeTypeCategory.TraitDefinition,
        NodeTypeCategory.InterfaceDefinition,
    ]);

    constructor(allNodeTypes: NodeTypeConfig[]) {
        this.statementsSuperSet = getQueryStatementsByCategories(
            allNodeTypes,
            ...this.nodeTypeCategories,
        );
        this.addQueriesForTSAndTSX();
        this.addQueriesForCAndCpp();
        this.addQueriesForJava();
        this.addQueriesForPHP();
    }

    addQueriesForCAndCpp() {
        this.statementsSuperSet.push(
            new SimpleLanguageSpecificQueryStatement(
                "(struct_specifier body: (field_declaration_list)) @struct_definition",
                ["cpp", "c"],
            ),
        );

        this.statementsSuperSet.push(
            new SimpleLanguageSpecificQueryStatement(
                "(union_specifier body: (field_declaration_list)) @union_definition",
                ["cpp", "c"],
            ),
        );

        this.statementsSuperSet.push(
            new SimpleLanguageSpecificQueryStatement(
                "(class_specifier body: (field_declaration_list)) @class_definition",
                ["cpp"],
            ),
        );

        this.statementsSuperSet.push(
            new SimpleLanguageSpecificQueryStatement(
                `(enum_specifier [
    base: (_)
    body: (enumerator_list)
]) @enum_with_body_or_type

(enum_specifier "enum"
    [
        "class"
        "struct"
    ]
    !base
    !body
) @scoped_enum_without_body_or_type
`,
                ["cpp"],
            ),
        );

        this.statementsSuperSet.push(
            new SimpleLanguageSpecificQueryStatement(
                "(enum_specifier body: (enumerator_list)) @enum_definition",
                ["c"],
            ),
        );
    }

    addQueriesForTSAndTSX() {
        this.statementsSuperSet.push(
            new SimpleLanguageSpecificQueryStatement(
                "(type_alias_declaration value: (object_type))",
                ["ts", "tsx"],
            ),
        );
    }
    addQueriesForJava() {
        this.statementsSuperSet.push(
            new SimpleLanguageSpecificQueryStatement("(object_creation_expression (class_body))", [
                "java",
            ]),
        );
    }
    addQueriesForPHP() {
        this.statementsSuperSet.push(
            new SimpleLanguageSpecificQueryStatement(`(object_creation_expression "new" "class")`, [
                "php",
            ]),
        );
    }
    async calculate(parsedFile: ParsedFile): Promise<MetricResult> {
        const { language, tree } = parsedFile;
        const queryBuilder = new QueryBuilder(language);
        queryBuilder.setStatements(this.statementsSuperSet);

        const query = queryBuilder.build();
        let matches: QueryMatch[] = [];
        if (query !== undefined) {
            matches = query.matches(tree.rootNode);
        }

        dlog(this.getName() + " - " + matches.length);

        return {
            metricName: this.getName(),
            metricValue: matches.length,
        };
    }

    getName(): string {
        return FileMetric.classes;
    }
}
