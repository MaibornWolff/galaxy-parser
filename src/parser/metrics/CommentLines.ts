import { QueryBuilder } from "../queries/QueryBuilder";
import { TreeParser } from "../helper/TreeParser";
import { ExpressionMetricMapping, QueryStatementInterface } from "../helper/Model";
import { getExpressionsByCategory, getQueryStatements } from "../helper/Helper";
import { Metric, MetricResult, ParseFile } from "./Metric";
import { QueryMatch, SyntaxNode } from "tree-sitter";
import { debuglog, DebugLoggerFunction } from "node:util";

let dlog: DebugLoggerFunction = debuglog("metric-gardener", (logger) => {
    dlog = logger;
});

export class CommentLines implements Metric {
    private readonly statementsSuperSet: QueryStatementInterface[] = [];
    private readonly commentExpressions: string[] = [];
    private excludedFileHeaderComments: SyntaxNode[] = [];

    constructor(allNodeTypes: ExpressionMetricMapping[]) {
        this.statementsSuperSet = getQueryStatements(allNodeTypes, this.getName());
        this.commentExpressions = getExpressionsByCategory(allNodeTypes, this.getName(), "comment");
    }

    calculate(parseFile: ParseFile): MetricResult {
        const tree = TreeParser.getParseTree(parseFile);
        this.excludedFileHeaderComments = this.getExcludedFileHeaderComments(
            tree.rootNode.children
        );

        const queryBuilder = new QueryBuilder(parseFile.fileExtension, tree);
        queryBuilder.setStatements(this.statementsSuperSet);

        const query = queryBuilder.build();
        let matches: QueryMatch[] = [];
        if (query !== undefined) {
            matches = query.matches(tree.rootNode);
        }

        const commentLines = matches.reduce((accumulator, match) => {
            const captureNode = match.captures[0].node;

            if (
                this.isFileHeaderComment(captureNode) ||
                this.isCommentFollowedByClass(captureNode)
            ) {
                return accumulator;
            }

            const commentLinesWithText = this.findTextLines(captureNode.text).length;
            const docBlockLines = this.findDocBlockLines(captureNode.text).length;

            return accumulator + commentLinesWithText - docBlockLines;
        }, 0);

        dlog(this.getName() + " - " + commentLines);

        return {
            metricName: this.getName(),
            metricValue: commentLines,
        };
    }

    private isCommentFollowedByClass(node: SyntaxNode) {
        // this is not language independent
        return node.nextSibling?.type.includes("class");
    }

    private findTextLines(text: string) {
        return text.split(/\r\n|\r|\n/g).filter((entry) => /[a-zA-Z0-9]+/g.test(entry));
    }

    private findDocBlockLines(text: string) {
        return text.split(/\r\n|\r|\n/g).filter((entry) => /^[ *]*@[a-z]+/g.test(entry));
    }

    private isFileHeaderComment(captureNode: SyntaxNode) {
        return this.excludedFileHeaderComments.some((excludedComment) => {
            return (
                excludedComment.startPosition.row === captureNode.startPosition.row &&
                excludedComment.startPosition.column === captureNode.startPosition.column &&
                excludedComment.endPosition.row === captureNode.endPosition.row &&
                excludedComment.endPosition.column === captureNode.endPosition.column
            );
        });
    }

    private getExcludedFileHeaderComments(rootNodeChildren): SyntaxNode[] {
        const excludedComments: SyntaxNode[] = [];
        const clonedChildren = [...rootNodeChildren];

        if (clonedChildren[0] && this.commentExpressions.includes(clonedChildren[0]?.type)) {
            // The very first elements that are comments are assumed to belong to a licence/copyright comment
            let hasFollowingComments = this.commentExpressions.includes(clonedChildren[0]?.type);

            while (hasFollowingComments) {
                excludedComments.push(clonedChildren.shift() as SyntaxNode);
                hasFollowingComments = this.commentExpressions.includes(clonedChildren[0]?.type);
            }
        }

        return excludedComments;
    }

    getName(): string {
        return "comment_lines";
    }
}
