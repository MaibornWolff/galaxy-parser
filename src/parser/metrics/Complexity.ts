import { QueryBuilder } from "../queries/QueryBuilder";
import {
    ExpressionMetricMapping,
    ExpressionQueryStatement,
    NodeTypeCategory,
    OperatorQueryStatement,
    QueryStatementInterface,
    SimpleQueryStatement,
    SimpleLanguageSpecificQueryStatement
} from "../helper/Model";
import { FileMetric, Metric, MetricResult, ParseFile } from "./Metric";
import { debuglog, DebugLoggerFunction } from "node:util";
import Parser, { QueryMatch } from "tree-sitter";
import { Languages } from "../helper/Languages";

let dlog: DebugLoggerFunction = debuglog("metric-gardener", (logger) => {
    dlog = logger;
});

export class Complexity implements Metric {
    private complexityStatementsSuperSet: QueryStatementInterface[] = [];

    constructor(allNodeTypes: ExpressionMetricMapping[]) {
        for (const expressionMapping of allNodeTypes) {
            if (expressionMapping.metrics.includes(this.getName())) {
                if (expressionMapping.type === "statement") {
                    const { expression, category, operator, activated_for_languages, languages } =
                        expressionMapping;

                    if (category === NodeTypeCategory.BinaryExpression && operator !== undefined) {
                        this.complexityStatementsSuperSet.push(
                            new OperatorQueryStatement(
                                category,
                                operator,
                                languages,
                                activated_for_languages
                            )
                        );
                    } else if (category === NodeTypeCategory.Other) {
                        this.complexityStatementsSuperSet.push(
                            new ExpressionQueryStatement(
                                expression,
                                languages,
                                activated_for_languages
                            )
                        );
                    }
                }
            }
        }
        this.createCaseQueryStatements(allNodeTypes);
    }

    createCaseQueryStatements(allNodeTypes: ExpressionMetricMapping[]) {
        /*
         * Determine for which languages there is an explicit default label node type in the grammar.
         * Handle case label node types differently based upon this.
         */
        const withDefaultLabelNodeType = new Set<string>();
        const caseNodeTypes = new Set<ExpressionMetricMapping>();

        for (const nodeType of allNodeTypes) {

            if (nodeType.category === NodeTypeCategory.CaseLabel && nodeType.metrics.includes(this.getName())) {
                caseNodeTypes.add(nodeType);
            } else if (nodeType.category === NodeTypeCategory.DefaultLabel) {
                for (const language of nodeType.languages) {
                    withDefaultLabelNodeType.add(language);
                }
            }
        }

        console.log("Language grammars with explicit default label node type: ");
        for (const langAbbr of withDefaultLabelNodeType) {
            console.log(langAbbr);
        }

        for (const caseNodeType of caseNodeTypes) {
            const haveDefaultNodeType: string[] = [];
            const haveNoDefaultNodeType: string[] = [];

            for (const languageAbbr of caseNodeType.languages) {
                if (withDefaultLabelNodeType.has(languageAbbr)) {
                    haveDefaultNodeType.push(languageAbbr);
                } else {
                    haveNoDefaultNodeType.push(languageAbbr);
                }
            }

            // For languages which have also a default node type, always count this case node type:
            if (haveDefaultNodeType.length > 0) {
                this.complexityStatementsSuperSet.push(
                    new ExpressionQueryStatement(
                        caseNodeType.expression,
                        haveDefaultNodeType,
                        caseNodeType.activated_for_languages
                    )
                );
            }

            // Special treatment for languages which have no explicit default node type:
            if (haveNoDefaultNodeType.length > 0) {
                if (caseNodeType.expression == "case_statement") {
                    // Special treatment for "case_statement" which can have more than one child,
                    // because it also has the content of the case block as child(s).
                    //
                    // In this case, a case label can be differentiated from a default label by
                    // checking if there is a child with the field name "value":
                    this.complexityStatementsSuperSet.push(
                        new SimpleLanguageSpecificQueryStatement(
                            "(case_statement value: _ ) @case_statement",
                            haveNoDefaultNodeType,
                            caseNodeType.activated_for_languages
                        )
                    );
                } else {
                    // Default treatment for case-label syntax nodes that are also used for default labels.
                    // Check if the node has a named child. If so, it is a case label and not a default label:
                    this.complexityStatementsSuperSet.push(
                        new SimpleLanguageSpecificQueryStatement(
                            "(" + caseNodeType.expression + " (_) @named_child) @" + caseNodeType.expression,
                            haveNoDefaultNodeType,
                            caseNodeType.activated_for_languages
                        )
                    );
                }
            }
        }
    }

    async calculate(parseFile: ParseFile, tree: Parser.Tree): Promise<MetricResult> {
        const queryBuilder = new QueryBuilder(parseFile, tree);

        if (parseFile.language === Languages.Java) {
            //add query for instance init block in Java
            queryBuilder.setStatements(
                this.complexityStatementsSuperSet.concat(
                    new SimpleQueryStatement("(class_body (block)) @initBlock")
                )
            );
        } else {
            queryBuilder.setStatements(this.complexityStatementsSuperSet);
        }

        const query = queryBuilder.build();
        let matches: QueryMatch[] = [];
        if (query !== undefined) {
            matches = query.matches(tree.rootNode);
        }

        dlog(this.getName() + " - " + matches.length);

        return {
            metricName: this.getName(),
            metricValue: matches.length
        };
    }

    getName(): string {
        return FileMetric.complexity;
    }
}
