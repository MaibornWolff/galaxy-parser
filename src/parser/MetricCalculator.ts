import { McCabeComplexity } from "./metrics/McCabeComplexity";
import { Functions } from "./metrics/Functions";
import { Classes } from "./metrics/Classes";
import fs from "fs";
import { LinesOfCode } from "./metrics/LinesOfCode";
import { CommentLines } from "./metrics/CommentLines";
import { RealLinesOfCode } from "./metrics/RealLinesOfCode";
import { ExpressionMetricMapping } from "./helper/Model";
import { Configuration } from "./Configuration";
import { Metric, MetricResult, ParseFile } from "./metrics/Metric";
import nodeTypesConfig from "./config/nodeTypesConfig.json";

/**
 * Arranges the calculation of all basic single file metrics on multiple files.
 */
export class MetricCalculator {
    private readonly fileMetrics: Metric[] = [];
    private config: Configuration;

    /**
     * Constructs a new instance of {@link MetricCalculator} for arranging the calculation of all
     * basic single file metrics on multiple files.
     * @param configuration Configuration that should be applied for this new instance.
     */
    constructor(configuration: Configuration) {
        this.config = configuration;
        const allNodeTypes: ExpressionMetricMapping[] =
            nodeTypesConfig as ExpressionMetricMapping[];

        this.fileMetrics = [
            new McCabeComplexity(allNodeTypes),
            new Functions(allNodeTypes),
            new Classes(allNodeTypes),
            new LinesOfCode(),
            new CommentLines(allNodeTypes),
            new RealLinesOfCode(allNodeTypes),
        ];
    }

    /**
     * Calculates all basic single file metrics on the specified files.
     * @param parseFiles Files for which the metrics should be calculated.
     * @return Map that relates the file path of each file to a map of the calculated metrics.
     */
    calculateMetrics(parseFiles: ParseFile[]) {
        const sourcesRoot = fs.realpathSync(this.config.sourcesPath);

        console.log(
            "\n\n",
            "----- Calculating metrics for " + sourcesRoot + " recursively -----",
            "\n\n"
        );
        console.log(" --- " + parseFiles.length + " files detected", "\n\n");

        const fileMetrics = new Map<string, Map<string, MetricResult>>();

        for (const parseFile of parseFiles) {
            const metricResults = new Map<string, MetricResult>();
            fileMetrics.set(parseFile.filePath, metricResults);

            console.log(" ------------ Parsing File " + parseFile.filePath + ":  ------------ ");

            for (const metric of this.fileMetrics) {
                const metricResult = metric.calculate(parseFile);
                metricResults.set(metricResult.metricName, metricResult);
            }
        }

        return fileMetrics;
    }
}
