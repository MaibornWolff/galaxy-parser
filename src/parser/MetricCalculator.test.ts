import { MetricCalculator } from "./MetricCalculator";
import {
    expectError,
    expectMetric,
    expectNoError,
    expectNoMetric,
    spyOnConsoleErrorNoOp,
} from "../../test/metric-end-results/TestHelper";
import {
    ErrorFile,
    FileMetric,
    FileMetricResults,
    ParsedFile,
    SourceFile,
    UnsupportedFile,
} from "./metrics/Metric";
import { FileType, Language, languageToGrammar } from "./helper/Language";
import Parser from "tree-sitter";
import { Classes } from "./metrics/Classes";
import { CommentLines } from "./metrics/CommentLines";
import { Complexity } from "./metrics/Complexity";
import { Functions } from "./metrics/Functions";
import { LinesOfCode } from "./metrics/LinesOfCode";
import { MaxNestingLevel } from "./metrics/MaxNestingLevel";
import { RealLinesOfCode } from "./metrics/RealLinesOfCode";
import { LinesOfCodeRawText } from "./metrics/LinesOfCodeRawText";

vi.mock("fs", () => ({
    promises: {
        readFile: vi.fn(),
    },
}));

function initiateSpies() {
    vi.spyOn(Classes.prototype, "calculate").mockResolvedValue({
        metricName: FileMetric.classes,
        metricValue: 1,
    });
    vi.spyOn(CommentLines.prototype, "calculate").mockResolvedValue({
        metricName: FileMetric.commentLines,
        metricValue: 2,
    });
    vi.spyOn(Complexity.prototype, "calculate").mockResolvedValue({
        metricName: FileMetric.complexity,
        metricValue: 3,
    });
    vi.spyOn(Functions.prototype, "calculate").mockResolvedValue({
        metricName: FileMetric.functions,
        metricValue: 4,
    });
    vi.spyOn(LinesOfCode.prototype, "calculate").mockResolvedValue({
        metricName: FileMetric.linesOfCode,
        metricValue: 5,
    });
    vi.spyOn(MaxNestingLevel.prototype, "calculate").mockResolvedValue({
        metricName: FileMetric.maxNestingLevel,
        metricValue: 6,
    });
    vi.spyOn(RealLinesOfCode.prototype, "calculate").mockResolvedValue({
        metricName: FileMetric.realLinesOfCode,
        metricValue: 7,
    });
    vi.spyOn(LinesOfCodeRawText, "calculate").mockResolvedValue({
        metricName: FileMetric.linesOfCode,
        metricValue: 8,
    });
}

function initiateErrorSpies() {
    vi.spyOn(Classes.prototype, "calculate").mockRejectedValue(
        new Error("something went wrong when calculating classes metric"),
    );
    vi.spyOn(CommentLines.prototype, "calculate").mockRejectedValue(
        new Error("something went wrong when calculating commentLines metric"),
    );
    vi.spyOn(Complexity.prototype, "calculate").mockRejectedValue(
        new Error("something went wrong when calculating complexity metric"),
    );
    vi.spyOn(Functions.prototype, "calculate").mockResolvedValue({
        metricName: FileMetric.functions,
        metricValue: 1,
    });
    vi.spyOn(LinesOfCode.prototype, "calculate").mockResolvedValue({
        metricName: FileMetric.linesOfCode,
        metricValue: 2,
    });
    vi.spyOn(MaxNestingLevel.prototype, "calculate").mockRejectedValue(
        new Error("something went wrong when calculating maxNestingLevel metric"),
    );
    vi.spyOn(RealLinesOfCode.prototype, "calculate").mockRejectedValue(
        new Error("something went wrong when calculating realLinesOfCode metric"),
    );
    vi.spyOn(LinesOfCodeRawText, "calculate").mockResolvedValue({
        metricName: FileMetric.linesOfCode,
        metricValue: 8,
    });
}

describe("MetricCalculator.calculateMetrics()", () => {
    let metricCalculator: MetricCalculator;
    let parsedFile: ParsedFile;
    let parsedFilePromise: Promise<SourceFile>;
    let parser: Parser;
    let tree: Parser.Tree;

    beforeAll(() => {
        metricCalculator = new MetricCalculator();
        parser = new Parser();
    });

    beforeEach(() => {
        // Clear all mock implementations, reset them to original implementation:
        vi.restoreAllMocks();
        spyOnConsoleErrorNoOp();
    });

    it("should calculate all metrics of type source code for a python file", async () => {
        // given
        parser.setLanguage(languageToGrammar.get(Language.Python));
        tree = parser.parse("sum(range(4))");
        parsedFile = new ParsedFile("test.py", Language.Python, tree);
        parsedFilePromise = new Promise<ParsedFile>((resolve) => {
            resolve(parsedFile);
        });

        initiateSpies();

        // when
        const [sourceFile, fileMetricResults] =
            await metricCalculator.calculateMetrics(parsedFilePromise);

        // then
        expect(sourceFile).toEqual(parsedFile);
        expect(fileMetricResults.fileType).toEqual(FileType.SourceCode);
        expectMetric(fileMetricResults, FileMetric.classes, 1);
        expectMetric(fileMetricResults, FileMetric.commentLines, 2);
        expectMetric(fileMetricResults, FileMetric.complexity, 3);
        expectMetric(fileMetricResults, FileMetric.functions, 4);
        expectMetric(fileMetricResults, FileMetric.linesOfCode, 5);
        expectMetric(fileMetricResults, FileMetric.realLinesOfCode, 7);
        expectNoMetric(fileMetricResults, FileMetric.maxNestingLevel);
    });

    it("should calculate lines of code and maximum nesting level for a JSON file", async () => {
        // given
        parser.setLanguage(languageToGrammar.get(Language.JSON));
        tree = parser.parse('{ "a": { "b": "c" } }');
        parsedFile = new ParsedFile("test.json", Language.JSON, tree);
        parsedFilePromise = new Promise<ParsedFile>((resolve) => {
            resolve(parsedFile);
        });

        initiateSpies();

        // when
        const [sourceFile, fileMetricResults] =
            await metricCalculator.calculateMetrics(parsedFilePromise);

        // then
        expect(sourceFile).toEqual(parsedFile);
        expect(fileMetricResults.fileType).toEqual(FileType.StructuredText);
        expectNoMetric(fileMetricResults, FileMetric.classes);
        expectNoMetric(fileMetricResults, FileMetric.commentLines);
        expectNoMetric(fileMetricResults, FileMetric.complexity);
        expectNoMetric(fileMetricResults, FileMetric.functions);
        expectMetric(fileMetricResults, FileMetric.linesOfCode, 5);
        expectNoMetric(fileMetricResults, FileMetric.realLinesOfCode);
        expectMetric(fileMetricResults, FileMetric.maxNestingLevel, 6);
    });

    it("should calculate lines of code for a text file", async () => {
        // given
        const unsupportedFile = new UnsupportedFile("test.txt");
        const unsupportedFilePromise = new Promise<UnsupportedFile>((resolve) => {
            resolve(unsupportedFile);
        });

        initiateSpies();

        // when
        const [sourceFile, fileMetricResults] =
            await metricCalculator.calculateMetrics(unsupportedFilePromise);

        // then
        expect(sourceFile).toEqual(unsupportedFile);
        expect(fileMetricResults.fileType).toEqual(FileType.Unsupported);
        expectNoMetric(fileMetricResults, FileMetric.classes);
        expectNoMetric(fileMetricResults, FileMetric.commentLines);
        expectNoMetric(fileMetricResults, FileMetric.complexity);
        expectNoMetric(fileMetricResults, FileMetric.functions);
        expectMetric(fileMetricResults, FileMetric.linesOfCode, 8);
        expectNoMetric(fileMetricResults, FileMetric.realLinesOfCode);
        expectNoMetric(fileMetricResults, FileMetric.maxNestingLevel);
    });

    it("should return an empty map of metrics when the source file is an error file", async () => {
        // given
        const errorFile = new ErrorFile(
            "/path/to/error/causing/file.js",
            new Error(
                "Root node of syntax tree for file /path/to/error/causing/file.js is undefined!",
            ),
        );

        parsedFilePromise = new Promise<SourceFile>((resolve) => {
            resolve(errorFile);
        });

        // when
        const result = await metricCalculator.calculateMetrics(parsedFilePromise);

        // then
        const expectedResult: [SourceFile, FileMetricResults] = [
            errorFile,
            { fileType: FileType.Error, metricResults: [], metricErrors: [] },
        ];
        expect(result).toEqual(expectedResult);
    });

    it("should include an error object in the result when an error is thrown while calculating any metric on a source file", async () => {
        // given
        parser.setLanguage(languageToGrammar.get(Language.Python));
        tree = parser.parse("sum(range(4))");
        parsedFile = new ParsedFile("test.py", Language.Python, tree);
        parsedFilePromise = new Promise<ParsedFile>((resolve) => {
            resolve(parsedFile);
        });

        initiateErrorSpies();

        // when
        const [sourceFile, fileMetricResults] =
            await metricCalculator.calculateMetrics(parsedFilePromise);

        // then
        expect(sourceFile).toEqual(parsedFile);
        expect(fileMetricResults.fileType).toEqual(FileType.SourceCode);
        expectNoMetric(fileMetricResults, FileMetric.classes);
        expectNoMetric(fileMetricResults, FileMetric.commentLines);
        expectNoMetric(fileMetricResults, FileMetric.complexity);
        expectMetric(fileMetricResults, FileMetric.functions, 1);
        expectMetric(fileMetricResults, FileMetric.linesOfCode, 2);
        expectNoMetric(fileMetricResults, FileMetric.realLinesOfCode);
        expectNoMetric(fileMetricResults, FileMetric.maxNestingLevel);

        expectError(
            fileMetricResults,
            FileMetric.classes,
            new Error("something went wrong when calculating classes metric"),
        );
        expectError(
            fileMetricResults,
            FileMetric.commentLines,
            new Error("something went wrong when calculating commentLines metric"),
        );
        expectError(
            fileMetricResults,
            FileMetric.complexity,
            new Error("something went wrong when calculating complexity metric"),
        );
        expectNoError(fileMetricResults, FileMetric.functions);
        expectNoError(fileMetricResults, FileMetric.linesOfCode);
        expectError(
            fileMetricResults,
            FileMetric.realLinesOfCode,
            new Error("something went wrong when calculating realLinesOfCode metric"),
        );

        // Should not have been tried to calculate on source code in the first place:
        expectNoMetric(fileMetricResults, FileMetric.maxNestingLevel);
    });
});
