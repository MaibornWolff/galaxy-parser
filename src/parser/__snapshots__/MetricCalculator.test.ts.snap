// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MetricCalculator.calculateMetrics() > should calculate all metrics of type source code for a python file 1`] = `
{
  "fileType": "source_code",
  "metricErrors": [],
  "metricResults": [
    {
      "metricName": "complexity",
      "metricValue": 3,
    },
    {
      "metricName": "functions",
      "metricValue": 4,
    },
    {
      "metricName": "classes",
      "metricValue": 1,
    },
    {
      "metricName": "lines_of_code",
      "metricValue": 5,
    },
    {
      "metricName": "comment_lines",
      "metricValue": 2,
    },
    {
      "metricName": "real_lines_of_code",
      "metricValue": 7,
    },
    {
      "metricName": "CommentKeywords",
      "metricValue": 0,
    },
  ],
}
`;

exports[`MetricCalculator.calculateMetrics() > should calculate lines of code and maximum nesting level for a JSON file 1`] = `
{
  "fileType": "structured_text",
  "metricErrors": [],
  "metricResults": [
    {
      "metricName": "lines_of_code",
      "metricValue": 5,
    },
    {
      "metricName": "max_nesting_level",
      "metricValue": 6,
    },
  ],
}
`;

exports[`MetricCalculator.calculateMetrics() > should calculate lines of code for a text file 1`] = `
{
  "fileType": "unsupported_file",
  "metricErrors": [],
  "metricResults": [
    {
      "metricName": "lines_of_code",
      "metricValue": 8,
    },
  ],
}
`;

exports[`MetricCalculator.calculateMetrics() > should include an error object in the result when an error is thrown while calculating any metric on a source file 1`] = `
{
  "fileType": "source_code",
  "metricErrors": [
    {
      "error": [Error: something went wrong when calculating complexity metric],
      "metricName": "complexity",
    },
    {
      "error": [Error: something went wrong when calculating classes metric],
      "metricName": "classes",
    },
    {
      "error": [Error: something went wrong when calculating commentLines metric],
      "metricName": "comment_lines",
    },
    {
      "error": [Error: something went wrong when calculating realLinesOfCode metric],
      "metricName": "real_lines_of_code",
    },
  ],
  "metricResults": [
    {
      "metricName": "functions",
      "metricValue": 1,
    },
    {
      "metricName": "lines_of_code",
      "metricValue": 2,
    },
    {
      "metricName": "CommentKeywords",
      "metricValue": 0,
    },
  ],
}
`;
